File: Objective-C Sample Code/Chapter 8/CMWrapper.m
//
//  CMWrapper.m
//  Top10
//
//  Created by Cory Bohon on 12/15/16.
//  Copyright © 2016 Cory Bohon. All rights reserved.
//

#import "CMWrapper.h"

@interface CMWrapper ()

@property (nonatomic, strong) NSOperationQueue *operationQueue;
@property (nonatomic, strong) CMMotionManager *motionManager;

@end

@implementation CMWrapper

+ (instancetype)sharedInstance
{
    static CMWrapper *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[self alloc] init];
        [sharedInstance setupWrapper];
    });
    
    return sharedInstance;
}

- (void)setupWrapper
{
    self.motionManager = [[CMMotionManager alloc] init];
    self.operationQueue = [[NSOperationQueue alloc] init];
}

- (void)setAccelerometerBlock:(void (^)(CMAccelerometerData *, NSError *))accelerometerBlock
{
    [self.motionManager setAccelerometerUpdateInterval:0.1f];
    [self.motionManager startAccelerometerUpdatesToQueue:self.operationQueue withHandler:accelerometerBlock];
}

- (void)stopMonitoring
{
    [self.motionManager stopAccelerometerUpdates];
    [self.motionManager stopDeviceMotionUpdates];
    [self.motionManager stopGyroUpdates];
    [self.motionManager stopMagnetometerUpdates];
}

@end

File: Swift Sample Code/Chapter 8/CMWrapper.swift
//
//  CMWrapper.swift
//  SampleApp
//
//  Created by Cory Bohon on 2/14/17.
//  Copyright © 2017 Cory Bohon. All rights reserved.
//

import UIKit
import CoreMotion

class CMWrapper: NSObject {

    static let shared = CMWrapper()
    var operationQueue: OperationQueue?
    var motionManager: CMMotionManager?
    
    override init() {
        super.init()
        self.setupWrapper()
    }

    func setupWrapper() {
        self.motionManager = CMMotionManager()
        self.operationQueue = OperationQueue()
    }
    
    func setAccelerometerBlock(accelerometerBlock: @escaping CMAccelerometerHandler) {
        self.motionManager?.accelerometerUpdateInterval = 0.1
        self.motionManager?.startAccelerometerUpdates(to: self.operationQueue!, withHandler: accelerometerBlock)
    }
    
    func stopMonitoring() {
        self.motionManager?.stopAccelerometerUpdates()
        self.motionManager?.stopDeviceMotionUpdates()
        self.motionManager?.stopGyroUpdates()
        self.motionManager?.stopMagnetometerUpdates()
    }
}
